### ISSUE TITLE
After the update from 5.3.0 to 5.3.1, all Twig templates are "not defined".

### ISSUE THREAD

**Issue by cmette** [contao/contao#6987]
### Affected version(s)

5.3.1

### Description

In my development environment under laragon 6.0, sf6.4.5, PHP 8.2.15, I have updated from C5.3.0 to 5.3.1 for testing purposes.
After this update, all Twig templates of the bundles under development are no longer accessible, although they are still in the corresponding folders.

I get the following exception: 

```
Template "@Contao_VendorBundleName/backend/DataContainerListener/MySpecialListener.list.label.label.twig" is not defined.
```

This applies to all 10 templates used by the bundle. 

Unfortunately, the following efforts do not bring any change:
- deleting the composer.lock and the vendor folder and run composer install
- delete the session
- delete the cache
- run composer dump-autoload
- rebuilding the symlinks

Could this be a bug?

**Comment by fritzmg** [comment]
Does it work in Contao 5.2.8 and 5.3.0?

**Comment by cmette** [comment]
I don't know if with 5.2.8. I started my bundle with 5.2 and have constantly updated it. From 5.2 up to and including 5.3.0 I had no problems...

**Comment by fritzmg** [comment]
> I don't know if with 5.2.8

Well, test it? 😉 This would help narrow down any possible regressions.

**Comment by cmette** [comment]
OK I have downgraded to 5.2.8. The exception is gone. Everything works as before without errors.

**Comment by fritzmg** [comment]
@m-vo may be a regression of https://github.com/contao/contao/pull/6936 ? 😕

**Comment by m-vo** [comment]
How can this be reproduced on a clean installation?

**Comment by cmette** [comment]
OK, I have upgraded to 5.3.0 again. The exception does not occur. Everything works as before without errors.

**Comment by eSpark-One** [comment]
After upgrading from 5.3.0 to 5.3.1 I got the same exception: **Twig templates are "not defined"**
`return $this->render('@Contao_VendorBundleName/content_element/be_grid.html.twig', $template->getData());`

I have downgraded to 5.3.0 and the error is gone.

**Comment by fritzmg** [comment]
> After upgrading from 5.3.0 to 5.3.1 I got the same exception: **Twig templates are "not defined"**

That is a different error message. You can seek help in the Contao Community (Slack or Forum for example).

**Comment by Tastaturberuf** [comment]
I have the same issue. In `4.13.37` everything works. Update to `4.13.38`:
```
Template "@Contao_App/Cards/CardBase.html.twig" is not defined in [@Contao_App/CardHearingAidSales.html.twig at line 1](https://intern.hoerpartner.local/preview.php/_profiler/open?file=@Contao_App/CardHearingAidSales.html.twig&line=1#line1).```

Files are under `contao/templates/Cards/CardBase.html.twig` and `contao/templates/Cards/CardHearingAidSales.html.twig`.

**Comment by Tastaturberuf** [comment]
I have the same issue. In `4.13.37` everything works. Update to `4.13.38`:
```
Template "@Contao_App/Cards/CardBase.html.twig" is not defined in [@Contao_App/CardHearingAidSales.html.twig at line 1].
```

Files are under `contao/templates/Cards/CardBase.html.twig` and `contao/templates/Cards/CardHearingAidSales.html.twig`.

**Comment by Tastaturberuf** [comment]
I think its part of #6949 

I will try to reproduce it later with the demo.

**Comment by fritzmg** [comment]
Where and  how are you rendering this template?

**Comment by Tastaturberuf** [comment]
With `insert()` in an Module-Template (.html5).

**Comment by fritzmg** [comment]
`$this->insert()` only supports HTML5 templates - at least there was never any official support for Twig templates this way.

@cmette how and where are you rendering your templates?

**Comment by cmette** [comment]
yes - it doesn't have to be a bug! hence my question: **could it be a bug?** 

first of all I apologize for my unconventional use of twig :blush: 

I use (somewhat carelessly) this context (schematic representation) - note the mixed use of $twig->render() and $this->twig->render()

where: **src/Controller/Backend/MyBackendController.php**
and 
how:

```php
<?php
class MyBackendController extends AbstractBackendController 
{
    #[Route('/import', defaults: ['_scope' => 'backend'])]
    public function import(): Response
    {
        // get a local twig instance to render large fragments 
       $twig = System::getContainer()->get('twig');

        // pre-render components
       $fieldsets = $twig->render(
           // depending on the context, this can be component1 or component2 etc.
           // I render here different formats like html, javascript, xml or xslt and so on 
           'a/special/path/to/my/component1/template.twig',
           [myTemplateData...]
       );

       // an so on...
       $submit_buttons = $twig->render(
           'a/special/path/to/my/button/components/button_set.twig',
           [myTemplateData...]
       );

        // render a copy of be_main in PageContent.twig
        return $this->render(
            "$this->templateRoot/PageContent.twig",
            [
                // header
                'headline'      => 'Eine Datei laden...',
                'message'       => $message,
                'buttons'       => $buttons,
                // form
                'form'          => $arrMainFormDef['form'],
                'fieldsets'     => $fieldsets,
                'submitButtons' => $submit_buttons,
            ]
        );
    }
}
```

**Comment by Tastaturberuf** [comment]
> $this->insert() only supports HTML5 templates - at least there was never any official support for Twig templates this way.

Is this written anywhere or is that an assertion?

At least up to version `4.13.37` this works fine with Twig templates.

**Comment by fritzmg** [comment]
@cmette please post the actual code - otherwise it is not possible to figure out the usage. At the very least we need the real template paths.

**Comment by cmette** [comment]
@fritzmg no problem - but unfortunately it will be tomorrow. the actual code? - that should be a lot... 
Would one controller be enough? At very least - the real paths are not static but are generated by functions... I'ill take a look tomorrow.

**Comment by m-vo** [comment]
Please also check and post the `debug:contao-twig <template>` output to verify the state of the loader.

**Comment by cmette** [comment]
First of all, I apologize for the late reply

under C3.5.0
template path: @Contao_CmetteContaoPedigreeBundle/backend/DataContainerListener/PedigreeImageListener.list.label.label.twig

**debug:twig @Contao_CmetteContaoPedigreeBundle**
output:
```
------------------------------------ --------------------------------------------------------------------------------------------
  Namespace                            Paths
 ------------------------------------ --------------------------------------------------------------------------------------------
  @Contao_CmetteContaoPedigreeBundle   F:/PHPStormProjects/contao-pedigree-bundle/contao/templates\
                                       F:/PHPStormProjects/contao-pedigree-bundle/contao/templates/backend\
                                       F:/PHPStormProjects/contao-pedigree-bundle/contao/templates/backend/DataContainerListener\
                                       F:/PHPStormProjects/contao-pedigree-bundle/contao/templates/backend/Pages\
                                       F:/PHPStormProjects/contao-pedigree-bundle/contao/templates/backend/ResumableWidget\
                                       F:/PHPStormProjects/contao-pedigree-bundle/contao/templates/modules\
 ------------------------------------ --------------------------------------------------------------------------------------------
```

after an update to 3.5.1

template path: @Contao_CmetteContaoPedigreeBundle/backend/DataContainerListener/PedigreeImageListener.list.label.label.twig

**debug:twig @Contao_CmetteContaoPedigreeBundle**
output:
```
Matched File
------------

 Template name
 "@Contao_CmetteContaoPedigreeBundle/backend/DataContainerListener/PedigreeImageListener.list.label.label.twig" not
 found

Configured Paths
----------------

 No template paths configured for "@Contao_CmetteContaoPedigreeBundle" namespace
```

Here is an example of a DCAListener:

```php
<?php

declare(strict_types=1);

/*
 * Contao Pedigree Bundle for Contao Open Source CMS
 *
 * Copyright (c) 2023 C. Mette
 *
 * @package    contao-pedigree-bundle
 * @link       https://github.com/cmette/contao-pedigree-bundle
 * @license    LGPL-3.0-or-later
 * @author     Christian Mette
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cmette\ContaoPedigreeBundle\EventListener\DataContainer;

use Cmette\ContaoPedigreeBundle\Models\PedigreeImageModel;
use Cmette\ContaoPedigreeBundle\Models\PedigreePeopleModel;
use Cmette\ContaoPedigreeBundle\Models\PedigreeSourceModel;

use Contao\BackendUser;
use Contao\Controller;
use Contao\CoreBundle\DependencyInjection\Attribute\AsCallback;
use Contao\DataContainer;
use Contao\Image;
use Contao\StringUtil;
use Contao\System;
use Contao\Message;

use Symfony\Component\HttpFoundation\Request;

class PedigreeImageListener extends PedigreeListener
{
    public const IMAGE_TEMPLATE = "<div class='%s' style='width:%s;height:%s;'>%s</div>";

    /**
     * a single person in a tag list
     */
    private const PERSON_TEMPLATE = '<span>%s<b>%s</b>%s</span>';

    public function __construct()
    {
        parent::__construct();

        $this->className = $this->getClassName();
        $this->arrLang   = $GLOBALS['TL_LANG']['tl_pedigree_image'];
    }

    #[AsCallback(table: 'tl_pedigree_image', target: 'config.onload')]
    private function onLoad(DataContainer $dc)
    {
        #dump($dc);
        #die();
    }
    /**********************************************************************
     * global_operations
     **********************************************************************/
    /**
     * @param string|null $href
     * @param string $label
     * @param string $title
     * @param string $class
     * @param string $attributes
     * @param string $table
     * @param array $rootRecordIDs
     * @return string
     */
    #[AsCallback(table: 'tl_pedigree_image', target: 'list.global_operations.scan.button')]
    public function listGlobalOperationsScanButton(?string $href, string $label, string $title, string $class, string $attributes, string $table, array $rootRecordIDs): string
    {
        // calls PedigreeImageController action scan
        return '<a href="/contao/image/scan" class="'.$class.'" title="'.StringUtil::specialchars($title).'"'.$attributes.'>'.$label.'</a> ';
    }

    /**
     * list label callback to display a single list row
     * handles different templates by action
     *
     * @param array         $row    record data
     * @param string        $label  current label
     * @param DataContainer $dc
     * @param array         $labels
     *
     * @return array
     */
    #[AsCallback(table: 'tl_pedigree_image', target: 'list.label.label')]
    public function ListLabelLabelCallback(array $row, string $label, DataContainer $dc, array $labels): string|array
    {
        /* @var Request $currentRequest */
        $currentRequest = System::getContainer()->get('request_stack')->getCurrentRequest();
        $strQueryString = $currentRequest->getQueryString();

        // select the template by action
        if(is_bool(strpos($strQueryString, "act=edit"))) {
            // render normal list.label.label
            $strTemplate= "$this->templateRoot/{$this->className}.list.label.label.twig";
            // label has no link
            $a          = [];
            // set the thumbnail size
            $thumbSize  = [160, 160, 'proportional'];
        } else {
            // render the picker list in edit-mode
            $strTemplate = "$this->templateRoot/{$this->className}.edit.picker.twig";
            // build link attributes
            $a['href']  = Controller::addToUrl("&id={$row['id']}&rt=" . $this->requestToken);
            $a['title'] = "Gehe zu Aufnahme mit der Signatur {$row['signature']}"; // ToDo: use TL_LANG
            // set the thumbnail size
            $thumbSize  = [80, 80, 'proportional'];
        }
        // build the images
        $frontSideImage = $this->buildPedigreeImage($row['frontside_singleSRC'], self::IMAGE_TEMPLATE, 'frame_frontside', $thumbSize);
        $backSideImage  = $this->buildPedigreeImage($row['backside_singleSRC'], self::IMAGE_TEMPLATE, 'frame_backside', $thumbSize);

        $people         = $this->renderPeopleTags($row['people_tags']);
        $comment        = $row['comment'];
        $source         = $this->getSourceById($row['source_id']);
        $source_type    = $this->arrLang['source_type_options'][$row['source_type']];
        $questions      = $row['questions'];
        $series         = $this->getImageLinkListFromSeries($row['series_ids']);

        $image          = $row;
        $image['signaturetitle'] = "Signatur: {$row['signature']} ID: {$row['id']} ";
dump($strTemplate);
        $html = $this->twig->render(
            $strTemplate,
            [
                'image'         => $image,
                'frontsideImage'=> $frontSideImage,
                'backsideImage' => $backSideImage,
                'a'             => $a,                  // array of <a>-tag-attributes
                'people'        => $people,
                'comment'       => $comment,
                'source'        => $source,
                'source_type'   => $source_type,
                'questions'     => $questions,
                'series'        => $series,
            ]
        );

        return $html;
    }

    /**********************************************************************
     * operations
     **********************************************************************/
    /**
     * @param array $row
     * @param $href
     * @param $label
     * @param $title
     * @param $icon
     * @param $attributes
     * @return string
     */
    #[AsCallback(table: 'tl_pedigree_image', target: 'list.operations.copy.button')]
    public function listCopyButton(array $row, $href, $label, $title, $icon, $attributes): string
    {
        if (!BackendUser::getInstance()->isAdmin) {
            $url        = '';
            $icon       = "copy--disabled.svg";
            $label      = $this->arrLang['copy_lock'];
            $_title     = StringUtil::specialchars($label);
            $entity     = $this->arrLang['singular'];
            $attributes = " class=\"copy\" data-action=\"contao--scroll-offset#store\" onclick=\"\"";
        } else {
            $url = Controller::addToUrl("$href&id={$row['id']}&rt=" . $this->requestToken);
            $_title = StringUtil::specialchars($title);
        }

        $image = Image::getHtml($icon, $label);

        return "<a href='$url' title='$_title' $attributes>$image</a>";
    }

    /**
     * controls the delete button depending on the use of the givenname
     *
     * @param $href
     * @param $label
     * @param $title
     * @param $icon
     * @param $attributes
     */
    #[AsCallback(table: 'tl_pedigree_image', target: 'list.operations.delete.button')]
    public function listDeleteButton(array $row, $href, $label, $title, $icon, $attributes): string
    {
        if (!BackendUser::getInstance()->isAdmin) {
            $url        = '';
            $icon       = "delete--disabled.svg";
            $label      = $this->arrLang['delete_lock'];
            $_title     = StringUtil::specialchars($label);
            $entity     = $this->arrLang['singular'];
            $attributes = " class=\"delete\" data-action=\"contao--scroll-offset#store\" onclick=\"\"";
        } else {
            $url = Controller::addToUrl("$href&id={$row['id']}&rt=" . $this->requestToken);
            $_title = StringUtil::specialchars($title);
        }

        $image = Image::getHtml($icon, $label);

        return "<a href='$url' title='$_title' $attributes>$image</a>";
    }


    //#[AsCallback(table: 'tl_pedigree_image', target: 'list.sorting.child_record')]
    private function childRecordCallback(array $row): string
    {
        return 'html';
    }

    /**
     * builds the people tag options for the field people_tags
     *
     * @param DataContainer $ds
     * @return array
     */
    #[AsCallback(table: 'tl_pedigree_image', target: 'fields.people_tags.options')]
    public function peopleTagsOptions(DataContainer $ds): array
    {
        $options = [];

        $people = PedigreePeopleModel::findAll();

        /* @var PedigreePeopleModel $person */
        if ($people) foreach ($people as $person) $options[$person->id] = $person->getFullNameAndLifetimeAsText('Y');

        return $options;
    }

    private function renderPeopleTags(?string $tags, string $className = ''): string
    {
        if (empty($tags)) {
            return 'keine Tags';
        }

        $arrTags = StringUtil::deserialize($tags);

        $arr = [];

        foreach ($arrTags as $personId) {
            $person = PedigreePeopleModel::findById($personId);
            if($person) {
                $arr[] = $person->getFullNameAsHtml();
            } else {
                $arr[] = "$personId?";
            }

        }

        return implode(' | ', $arr);
    }

    /**
     * if the series_ids has been updated,
     *  also update all data records of the referenced images
     *
     * @param mixed $value
     * @param DataContainer $dc
     * @return mixed
     */
    #[AsCallback(table: 'tl_pedigree_image', target: 'fields.series_ids.save')]
    public function fieldsSeriesIdsSave(mixed $value, DataContainer $dc): mixed
    {
        if($value) {
            // at least one further image is available
            $ids    = StringUtil::deserialize($value);
            $ids[]  = $dc->id;

            $otherImages = PedigreeImageModel::findMultipleByIds($ids);

            if($otherImages) {
                /* @var PedigreeImageModel $image */
                foreach ($otherImages as $image) {
                    $image->updateSeriesIds($ids);
                }
            } else {
                Message::addError('Die referenzierten Aufnahmen sind nicht vorhanden.'); // ToDo: TL_LANG
            }
        }
        return $value;
    }

    private function getSourceById($id): string
    {
        $result = '?';

        if($source = PedigreeSourceModel::findById($id)) {
            $result = $source->name;
        };

        return $result;
    }

}

```

**Comment by m-vo** [comment]
> debug:twig @Contao_CmetteContaoPedigreeBundle

You are confusing `debug:twig` (which shows information regarding Symfony's filesystem loader with `debug:contao-twig` (which is about Contao's loader and the managed namespace system). You cannot expect Symfonys loader to find templates in Contao template paths and vice versa.

Please read about how templating works in Contao, which is all documented here: https://docs.contao.org/dev/framework/templates/

You might also want to reconsider your naming and usage standards which looks very strange to me (and most likely others) if this is a public bundle/part of the public Contao ecosystem. :wink:

**Comment by m-vo** [comment]
**@&ZeroWidthSpace;all:** The exception message `Template "<template>" is not defined.` originates from the `ChainLoader`, which means **none** of the registered loaders could find it, i.e. neither Symfony's nor ours. If you get a message like this, first find out which loader you *expect* to find the template, then check the spelling/correct namespace/… and use the appropriate debug command to double check. 

For the Contao side: if the template (the actual file!) is not reported by the `debug:contao-twig` command, it also does not exist for the loader. If it does not exist, make sure it's in a Contao template directory and that this directory has a `.twig-root` file (if not implicitly present). Make sure the issue is not related to caching. It is expected, that you need to clear the cache when working in the prod environment. Currently, you might also need to clear the cache when creating templates in the backend (this will change in the future). 

If there is still an issue, please feel free to open up a new issue, that contains at least the following:

0) your Contao version
1) the exact location where you template file is located
2) if there is a `.twig-root` file in the directory or any parent directory
3) if in a bundle: how your bundle is named
4) your exact render call containing the real logical template name, that you are rendering - or, if you are using an abstract fragment controller: how the controller attribute/tag looks like and how it is named.

This issue got convoluted with a lot of different things, I'm closing this here. As said before, please feel free to open up new issues for specific problems.

**Comment by cmette** [comment]
but the fundamental question remains unanswered: Why does the error not occur under 5.3.0, but only after an update to 5.3.1?

**Comment by cmette** [comment]
@m-vo you write: You might also want to reconsider your naming and usage standards which looks very strange to me (and most likely others) if this is a public bundle/part of the public Contao ecosystem. 😉

That is very general. Please go into more detail. I use the templates from docs.contao.org insofar as they are meaningful at all.
(see: #https://docs.contao.org/dev/guides/back-end-routes/#create-your-controller-and-template)

**Comment by fritzmg** [comment]
> use the templates from docs.contao.org insofar as they are meaningful at all.
> (see: #https://docs.contao.org/dev/guides/back-end-routes/#create-your-controller-and-template)

`my_backend_route.html.twig` would live under `templates/my_backend_route.html.twig`, as explained in the docs.

**Comment by m-vo** [comment]
Your example points to a Symfony bundle template, please read the docs about the Contao templating system and or look at how the core implements templates: https://github.com/contao/contao/tree/5.x/core-bundle/contao/templates/twig (you don't need the `twig` subdirectory, though - that's only to group them away from the legacy stuff).

**Comment by cmette** [comment]
OK. I don't see any difference or the problem yet - but we don't need to discuss it here. I will go through the documentation again.
Many thanks for the tips!
